packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js

```diff
@@ -62,22 +65,37 @@ export default createConnector({
       results.queryID
     );
 
-    if (page === 0) {
-      this._allResults = hitsWithPositionsAndQueryID;
-    } else if (page > this._previousPage) {
+    if (this._firstReceivedPage === undefined) {
+      this._allResults = [...hitsWithPositionsAndQueryID];
+      this._firstReceivedPage = page;
+      this._lastReceivedPage = page;
+    } else if (this._lastReceivedPage < page) {
       this._allResults = [...this._allResults, ...hitsWithPositionsAndQueryID];
-    } else if (page < this._previousPage) {
-      this._allResults = hitsWithPositionsAndQueryID;
+      this._lastReceivedPage = page;
+    } else if (this._firstReceivedPage > page) {
+      this._allResults = [...hitsWithPositionsAndQueryID, ...this._allResults];
+      this._firstReceivedPage = page;
     }
 
+    const hasPrevious = this._firstReceivedPage > 0;
     const lastPageIndex = nbPages - 1;
     const hasMore = page < lastPageIndex;
-
-    this._previousPage = page;
+    const refine = index => {
```

Comment:
[Same question](https://github.com/algolia/vue-instantsearch/pull/659#discussion_r274967396) by @Haroenv on the Vue implementation.

Copy-past of my response:
> Talked about it with @samouss during the RFC review ([discussion](https://github.com/algolia/instantsearch-rfcs/pull/16#discussion_r270315733)). I had `refine('previous')` and `refine('next')` but we agreed for `refineNext()` and `refinePrevious()`.
>
> But we can re-discuss it together :wink: 

If we decide to change it it should not be too much work, but it should be done in every flavor to stay consistent.

Reply:
Sorry, my question was different. It was about reusing the [`refine` function hook](https://www.algolia.com/doc/guides/building-search-ui/widgets/create-your-own-widgets/react#lifecyle-and-api) from the `createConnector` object instead of returning it in `getProvidedProps`.

It might make the code more complex though.

Reply:
Oh, my bad, I see what you means now. I'll try this.